tests/integration/api/controllers/contact.spec.js

```diff
@@ -0,0 +1,426 @@
+  const contact2 = utils.deepFreeze(personFactory.build({ name: 'contact2', role: 'program_officer' }));
+  const placeMap = utils.deepFreeze(placeFactory.generateHierarchy());
+  const place1 = utils.deepFreeze({ ...placeMap.get('health_center'), contact: { _id: contact1._id } });
+  const place2 = utils.deepFreeze({ ...placeMap.get('district_hospital'), contact: { _id: contact2._id } });
+  const place0 = utils.deepFreeze({
+    ...placeMap.get('clinic'),
+    contact: { _id: contact0._id },
+    parent: {
+      _id: place1._id,
+      parent: {
+        _id: place2._id
+      }
+    },
+  });
+  const patient = utils.deepFreeze(personFactory.build({
+    parent: {
+      _id: place0._id,
+      parent: {
+        _id: place1._id,
+        parent: {
+          _id: place2._id
+        }
+      },
+    },
+    phone: '1234567890',
+    role: 'patient',
+    short_name: 'Mary'
+  }));
+  const placeType = 'clinic';
+  const clinic1 = utils.deepFreeze(placeFactory.place().build({
+    parent: {
+      _id: place1._id,
+      parent: {
+        _id: place2._id
+      }
+    },
+    type: placeType,
+    contact: {},
+    name: 'clinic1'
+  }));
+  const clinic2 = utils.deepFreeze(placeFactory.place().build({
+    parent: {
+      _id: place1._id,
+      parent: {
+        _id: place2._id
+      }
+    },
+    type: placeType,
+    contact: {},
+    name: 'clinic2'
+  }));
+
+  const userNoPerms = utils.deepFreeze(userFactory.build({
+    username: 'online-no-perms',
+    place: place1._id,
+    contact: {
+      _id: 'fixture:user:online-no-perms',
+      name: 'Online User',
+    },
+    roles: ['mm-online']
+  }));
+  const offlineUser = utils.deepFreeze(userFactory.build({
+    username: 'offline-has-perms',
+    place: place0._id,
+    contact: {
+      _id: 'fixture:user:offline-has-perms',
+      name: 'Offline User',
+    },
+    roles: ['chw']
+  }));
+  const allDocItems = [contact0, contact1, contact2, place0, place1, place2, clinic1, clinic2, patient];
+  const dataContext = getRemoteDataContext(utils.getOrigin());
+  const personType = 'person';
+  const e2eTestUser = {
+    '_id': 'e2e_contact_test_id',
+    'type': personType,
+  };
+  const onlineUserPlaceHierarchy = {
+    parent: {
+      _id: place1._id,
+      parent: {
+        _id: place2._id,
+      }
+    }
+  };
+  const offlineUserPlaceHierarchy = {
+    parent: {
+      _id: place0._id,
+      ...onlineUserPlaceHierarchy
+    }
+  };
+  const expectedPeople = [
+    contact0,
+    contact1,
+    contact2,
+    patient,
+    e2eTestUser,
+    {
+      type: personType,
+      ...userNoPerms.contact,
+      ...onlineUserPlaceHierarchy
+    },
+    {
+      type: personType,
+      ...offlineUser.contact,
+      ...offlineUserPlaceHierarchy
+    }
+  ];
+  const expectedPeopleIds = expectedPeople.map(person => person._id);
+  const expectedPlaces = [place0, clinic1, clinic2];
+  const expectedPlacesIds = expectedPlaces.map(place => place._id);
+
+  before(async () => {
```

Comment:
All these tests should call the API endpoints, instead of requesting datasource code.

Reply:
This was heavily discussed with QA when laying down the original pattern for these tests.  The summary of the the IRL convos is here: https://github.com/medic/cht-core/pull/9090#discussion_r1613969729

TLDR is we _could_ duplicate tests, but the value of that seems nearly non-existant (while adding a non-zero cost in terms of additional tests to run).

Reply:
@sugat009 How feasible would it be to parameterize the tests where we are currently using cht-datasource so that we run the same test with _both_ the cht-datasource remote wrapper and with the test utils making the same REST call?

I am thinking of something like this (but have not actually tried to run this yet):
```js
    [
      (placeId) => Place.v1.get(dataContext)(Qualifier.byUuid(placeId)),
      (placeId) => {
        const opts = {
          path: `/api/v1/place/${placeId}`,
        };
        return utils.request(opts);
      }
    ].forEach((getPlace) => {
      it('returns the place matching the provided UUID', async () => {
        const place = await getPlace(place0._id);
        expect(place).excluding(['_rev', 'reported_date']).to.deep.equal(place0);
      });
    });
```

Reply:
That would ensure we test both the cht-datasource Remote adapter as well as just using the test utils to hit the REST api "directly" without actually needing to duplicate every test...